Split the registry into registry, web, and storage services

chrisdickinson opened the pull request:
Status: it's ready!

chrisdickinson commented on services/common/boltzmann/client/NOTES:
```diff
@@ -0,0 +1,100 @@
+storageApi
```
Hey look, it's almost a complete list of everything an entropic can do!

chrisdickinson commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+'use strict'
```
Both the web gateway and registry gateway can use this client to talk to storage.

chrisdickinson commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+  }) {
+    this.url = url
+    this.requestId = requestId
+    this.userAgent = `${hostname}(${userAgent})`
```
Wrap up the incoming requestId and userAgent (so we know which host is sending which user-agent where, in logs.)

Should I add `x-forwarded-for` here as well, so remote address logging works correctly in storage-api?

chrisdickinson commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+      body: rawBody ? rawBody : body ? JSON.stringify(body) : null
+    })
+
+    if (!response.ok) {
```
The fetch api is so nice.

chrisdickinson commented on services/common/boltzmann/request-handler.js:
```diff
@@ -22,6 +24,15 @@ class Context {
+      return this._parsedUrl
+    }
+    this._parsedUrl = parse(this.request.url, true)
+    return this._parsedUrl
```
Be lazy about URL parsing.

chrisdickinson commented on services/registry/handlers/maintainers.js:
```diff
@@ -16,88 +10,72 @@ module.exports = [
+      'maintainer.invite.package_dne': `Unknown package: "${invitee}".`,
+      'maintainer.invite.already_accepted': `Namespace "${invitee}" is already a member.`,
+      'maintainer.invite.already_declined': `Namespace "${invitee}" has declined this invite.`
+    }[err.code]
```
The error mapping pattern emerges. It's almost worth making a decorator for this and throwing the error, I think. This is where we should hang i18n mapping as well, potentially.

chrisdickinson commented on services/registry/middleware/bearer-auth.js:
```diff
@@ -34,16 +21,29 @@ function createBearerAuthMW() {
+      const hash = crypto
+        .createHash('sha256')
+        .update(bearer + process.env.SESSION_SECRET)
+        .digest('base64')
```
The registry now caches auth in redis for a span of time, so we hash it.

chrisdickinson commented on services/registry/package.json:
```diff
@@ -12,13 +12,10 @@
-    "find-my-way": "~2.0.1",
     "graceful-fs": "^4.1.15",
     "is-email-maybe": "^1.0.1",
     "markdown": "^0.5.0",
```
TODO: review all service deps and shrink to the minimum set for each service.

chrisdickinson replied:
TODO another day, perhaps

chrisdickinson commented:
This turned out to be much bigger than expected! (It's good to get this out of the way now, though, because it's just going to move _more_ cheese as time goes on.)

Things left to do:

- [ ] Finish wiring up the client
- [ ] Fill out the remaining "TODO" api endpoints from the client (things we were previously glomming together with the registry/web endpoints before, like token creation)
- [ ] Do a error enumeration pass through:
  - [ ] the registry
  - [ ] the website
  - [ ] the storage layer
- [ ] Do a logging pass through:
  - [ ] the registry
  - [ ] the website
  - [ ] the storage layer

olingern commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+'use strict'
+
+const fetch = require('node-fetch')
```
You might want to use the wrapped `fetch` that was merged in #177. #183 is still open that aims to do some refactoring of the original merge

chrisdickinson replied:
Ah, this is on the server side!

olingern commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+
+const hostname = os.hostname()
+
+const e = encodeURIComponent
```
What do you think of wrapping this in a function? If any future refactors need to happen it could make it easier.

`const encode = (str) => encodeURIComponent(str)`

olingern commented on services/common/boltzmann/client/index.js:
```diff
@@ -0,0 +1,207 @@
+
+module.exports = class Client {
+  constructor ({
+    url = process.env.STORAGE_API_URL,
```
Should we `assert`, `throw`, or warn here if that env var isn't defined?

olingern commented on services/web/server.js:
```diff
@@ -0,0 +1,33 @@
+} else {
+  bole.output({ level: 'info', stream: process.stdout });
+}
+
```
My go at code golfing the wrapper:

**Wrapper**
```javascript
function boltzStarter(router, middles, boltzmann) {
  return boltzmann.make(router, middles.map(middle => require(middle)));
}
```
**Usage**
```javascript
const myMiddles = [
  'boltzmann/middleware/logger',
  'boltzmann/middleware/flush-request',
  'boltzmann/middleware/requestid',
  'boltzmann/middleware/redis',
  'boltzmann/middleware/storage-api',
  './middleware/session',
];

boltzStarter(router, myMiddles, boltzmann).listen(
  process.env.PORT,
  '0.0.0.0'
);
```

zacanger replied:
Oh, I commented above but it looks like y'all are already on it.

zacanger replied:
I made a ticket for that [here](https://github.com/entropic-dev/entropic/issues/213) so it doesn't get lost

olingern commented on services/common/boltzmann/client/index.js:
```diff
@@ -185,9 +300,10 @@ module.exports = class Client {
         message = parsed.message || body
         code = parsed.code || code
       } catch {
+        ;
```
I'm sure you're probably working on this now, but commenting just in case. Empty line / no caught error

olingern commented on services/storage/handlers/namespaces.js:
```diff
@@ -0,0 +1,253 @@
+        'namespace_members.active': true,
+        'namespace_members.user_id': context.user.id
+      })
+      .catch(Namespace.objects.NotFound, () => null);
```
Could we add a log here? #202 took me a bit to track down :sweat_smile:

chrisdickinson replied:
This _should_ appear a lot noisier in the logs now after adding dev-mode traces to error responses.

olingern commented:
Awesome work! I thought I would try to help PR review and leave some comments. I'll try to test this branch out at some point to see if I can turn up any :beetle:

zacanger commented on services/common/boltzmann/request-handler.js:
```diff
@@ -1,5 +1,7 @@
 'use strict';
 
+const { parse } = require('url');
```
`url.parse` is deprecated, should this use `new URL` instead?

chrisdickinson replied:
Ah yeah. `new URL` has the (uncomfortable in this case) prereq that you have to give it two args to prevent it from throwing if the first arg is an "incomplete" url – we _probably_ should move over to it, but as a follow up PR.

zacanger commented on services/registry/handlers/maintainers.js:
```diff
@@ -16,88 +10,71 @@ module.exports = [
+    page: Number(context.url.query.page) || 0,
+    status: context.url.query.status
+  }).then(
+    xs => [null, xs],
```
I like this. Haskell influence?

chrisdickinson replied:
Haha, thanks! I got it from looking at Golang a while back, tbh.

zacanger commented on services/registry/handlers/users.js:
```diff
@@ -0,0 +1,94 @@
+    namespace,
+    host
+  }).then(
+    xs => [null, xs],
```
Would it be worth pulling this into an `either` util?

chrisdickinson replied:
That would be great once this lands!

zacanger commented on services/storage/handlers/packages.js:
```diff
@@ -0,0 +1,509 @@
+'use strict';
+
+const { Response } = require('node-fetch');
+const { markdown } = require('markdown');
```
This particular markdown lib is very abandoned. It might make sense to use something like markdown-it, which is both maintained and super pluggable.

chrisdickinson replied:
I was waiting to use `marky-markdown`, which is a Known Good package for rendering markdown; _however_ it brings along oniguruma so I wanted to wait until we had separated package publishes out into a worker process.

zacanger replied:
Added [a ticket](https://github.com/entropic-dev/entropic/issues/214) so this doesn't get lost.

zacanger commented on services/storage/handlers/packages.js:
```diff
@@ -0,0 +1,509 @@
+      './' + decodeURIComponent(String(part.filename)).replace(/^\/+/g, '');
+    formdata.files[filename] = context.storage.add(part);
+
+    if (/^\.\/package\/readme(\.(md|markdown))?/i.test(filename)) {
```
`.mkd` is also somewhat common

chrisdickinson replied:
great catch!

zacanger commented on services/storage/handlers/packages.js:
```diff
@@ -0,0 +1,509 @@
+  }
+
+  await Promise.all(
+    Object.keys(formdata.files).map(filename => {
```
`Object.values` might reduce some code here

chrisdickinson replied:
Ah, this is a particularly gross bit of code where I'm overwriting the key in the original object with the result of the promise, so I wouldn't be able to use `Object.values` here (since I need the key to thunk out the attribute.)

zacanger commented on services/storage/server.js:
```diff
@@ -0,0 +1,48 @@
+
+require('dotenv').config();
+
+const boltzmann = require('boltzmann');
```
I'd be happy to put in a PR to pull out some of this boilerplate, it'll probably become pretty common.

zacanger commented on services/web/handlers/index.js:
```diff
@@ -0,0 +1,31 @@
+    <!doctype html>
+    <html>
+      <body>
+        <h1><marquee>WELCOME TO ENTROPIC</marquee></h1>
```
`WELCOME TO 1997`

chrisdickinson replied:
It's a package registry on the move!

chrisdickinson commented:
OK, I'm calling it – this PR is ready to merge, I think. Brace yourself: I anticipate that it's going to break stuff and things will be a bit bumpy for a little while.

**Here's what I've done:**

- www.js has been split out of registry into `services/web`.
- Most of the registry code has moved back up into a new (private) service called `storage`.
    - The intent is that this will not be visible to the outside world.
- `web` and `registry` are responsible for trading external authentication (session cookies and authorization headers, respectively) for internal authentication (the surprisingly versatile `bearer: <username>` header.)
- `registry` caches authentication results in redis.
- membership (of users to namespaces) and maintainership (of packages by namespaces) URLs have been restructed **but the CLI has not caught up.** See **URL CHANGES** below.
- `docker-compose.yml` contains all the new services.
- Added `response.error.coded` which accepts a string (a la `namespace.members.bearer_unauthed`), designed to create a disjoint between `storage` layer error cases and `registry` layer error message display. (This means we can start pushing `i18n` concerns into the registry layer!)

**I've tested:**

- logging in both via services running locally and through docker
- publishing a new package
- testing `whoami`
- manually curling a few list endpoints – packages, maintainerships, memberships.

**I have not:**

- updated the documentation (for another PR!)
- added registry tests (but wow we need them! all of the existing tests – save invitations, which were deleted due to _massive upheaval_ – moved from registry to storage)
- audited the dependencies of each service. It's possible (likely!) that services are depending on things they aren't using.

## URL CHANGES

```
GET     /v1/users/user/:user/memberships[?pending=true]                list membership
POST    /v1/users/user/:user/memberships/invitations/:namespace@:host  accept invitation
DELETE  /v1/users/user/:user/memberships/invitations/:namespace@:host  decline invitatation

GET     /v1/namespaces/namespace/:namespace@:host/maintainerships[?pending=true]            list maintainerships
POST    /v1/namespaces/namespace/:namespace@:host/maintainerships/:namespace@:host/:package accept invite
DELETE  /v1/namespaces/namespace/:namespace@:host/maintainerships/:namespace@:host/:package decline invite

GET     /v1/namespaces/namespace/:namespace@:host/members[?pending=true] list members
POST    /v1/namespaces/namespace/:namespace@:host/members/:user          invite user
DELETE  /v1/namespaces/namespace/:namespace@:host/members/:user          remove member / rescind invite

GET     /v1/packages/package/:namespace@:host/maintainers[?pending=true] list maintainers
POST    /v1/packages/package/:namespace@:host/maintainers/:name@:host    invite maintainer
DELETE  /v1/packages/package/:namespace@:host/maintainers/:name@:host    remove maintainer / rescind invite
```

This eliminates all urls with the word `invite` in them, because we don't so much have invites as "pending" maintainerships and memberships; and the user/namespace is free to leave if they choose to do so. The CLI should be updated to reflect these changes! This can be done in a subsequent PR.

## why is it important to move all this cheese now

Moving this cheese:

- Gives people a service in which to build _all_ the website functionality in a connected way. (If we end up going with Next.js, we may need _yet another_ service that lives outside of the private network and uses `web` as an API gateway/backend-for-frontend. That's okay, though – we've already gone from 1 service to 3. 3 to 4 will be much easier.)
- Lets us separate expensive processes (like publishing and syncing!) into a worker process.

chrisdickinson commented on package.json:
```diff
@@ -30,8 +30,8 @@
+    "postinstall": "for d in cli services/{registry,workers,web,common/boltzmann}; do cd $d; npm i; cd -; done",
     "start": "docker-compose up",
-    "test": "for d in cli services/registry; do cd $d; npm t; cd -; done"
+    "test": "for d in cli services/{registry,workers,web,common/boltzmann}; do cd $d; npm t; cd -; done"
```
Use bash in anger.

chrisdickinson commented on services/common/boltzmann/client/NOTES:
```diff
@@ -0,0 +1,100 @@
+    getObject({algo, digest}) // promise for a stream
+
+
+response.rawjson(str) // for enemies
```
Leaving this for now, will clean up afterward. Sectumsempra!

chrisdickinson commented on services/common/boltzmann/response.js:
```diff
@@ -75,18 +77,32 @@ function error(err, status = 500, extraHeaders = {}) {
+
+  if (isDev()) {
+    err.trace = new Error().stack;
+  }
```
If we're in dev mode, add a `trace` property to all outgoing errors. This makes it easier to track down errors when they happen: the registry client will log traces coming from `storage`.

chrisdickinson commented on services/common/boltzmann/response.js:
```diff
@@ -75,18 +77,32 @@ function error(err, status = 500, extraHeaders = {}) {
 
+error.coded = coded;
+function coded(code, ...args) {
+  return error(Object.assign(new Error(code), { code }), ...args);
```
For storage-layer use. Takes computer-friendly dotted, lowercase error codes.

chrisdickinson commented on services/common/boltzmann/response.js:
```diff
@@ -75,18 +77,32 @@ function error(err, status = 500, extraHeaders = {}) {
   });
   if (typeof message === 'string') {
-    message = { error: message };
+    message = { message, code: 'authneeded' };
```
Standardize a bit on `message` (which is used in `response.error`), while we're at it.

chrisdickinson commented on services/registry/handlers/maintainers.js:
```diff
@@ -16,88 +10,70 @@ module.exports = [
-  fork.del(
-    '/v1/packages/package/:namespace([^@]+)@:host/:name/invitation/:member',
-    packageExists(isNamespaceMember(decline))
+    authn.required(remove)
```
A theme emerges: registry endpoints can make authorization decisions on a large scale – "this will fail if you aren't authenticated" – but delegate granular authorization decisions to the storage layer ("can Alice write to Bob's package?")

chrisdickinson commented on services/registry/handlers/maintainers.js:
```diff
@@ -16,88 +10,70 @@ module.exports = [
+      'maintainer.invite.package_dne': `Unknown package: "${invitee}".`,
+      'maintainer.invite.already_accepted': `Namespace "${invitee}" is already a member.`,
+      'maintainer.invite.already_declined': `Namespace "${invitee}" has declined this invite.`
+    }[err.code];
```
This is (more or less) what I'm thinking error enumeration should look like, per those many TODO comments.

chrisdickinson commented on services/registry/middleware/bearer-auth.js:
```diff
@@ -1,26 +1,15 @@
-function createBearerAuthMW() {
+function createBearerAuthMW({
+  sessionTimeout = Number(process.env.SESSION_TIMEOUT) || 5 * 60
+} = {}) {
```
The registry layer _always_ attempts to trade the `Authorization` header for a `User` object when given. It'll try redis first, then fall back to asking the storage layer.

chrisdickinson commented on services/registry/middleware/bearer-auth.js:
```diff
@@ -34,16 +23,34 @@ function createBearerAuthMW() {
+
+      if (data === null) {
+        const [err, result] = await context.storageApi
+          .getToken(bearer)
```
`getToken()` sends the token in a `{token}` header so sensitive data does not get logged. 🤫

chrisdickinson commented on services/storage/decorators/authn.js:
```diff
@@ -0,0 +1,30 @@
+'use strict';
```
The storage layer, however, only tries to load the User from the database if a handler _specifically requests it_ using one of these decorators.

chrisdickinson commented on services/storage/decorators/package-exists.js:
```diff
@@ -0,0 +1,41 @@
+  const { namespace, host, name } = mapping;
+  return next => {
+    return async (context, params) => {
+      const { [namespace]: ns, [host]: h, [name]: pkg } = params;
```
ES2015: the idents are dynamic, and the points are all free style

chrisdickinson commented on services/storage/handlers/index.js:
```diff
@@ -0,0 +1,37 @@
+'use strict';
```
Most of the `storage/handlers` are ports of the corresponding registry files.

chrisdickinson commented on services/storage/handlers/users.js:
```diff
@@ -0,0 +1,233 @@
+  return response.json({
+    objects,
+    next: hasNext,
+    prev: hasPrev
```
A pagination scheme appears.

chrisdickinson commented on services/storage/middleware/internal-auth.js:
```diff
@@ -0,0 +1,38 @@
+      // `context.user` return a promise for the currently
+      // authenticated user (if any.)
+      //
+      // TODO: Someday use JWTs, maybe.
```
Maybe.

chrisdickinson commented on services/storage/models/authentication.js:
```diff
@@ -1,60 +1,10 @@
-    );
-  }
-}
-
```
Provider data lives in `web` now.

chrisdickinson commented on services/storage/models/authentication.js:
```diff
@@ -96,6 +59,8 @@ module.exports = class Authentication {
   }
 };
 
+const User = require('./user');
```
Had to shuffle requires because of 🎶 circular imports :notes:

chrisdickinson commented on services/storage/models/token.js:
```diff
@@ -34,6 +34,11 @@ module.exports = class Token {
   }
 
+  toJSON() {
+    const { user: _0, id: _1, user_id: _2, ...meta } = this;
```
Omitting properties by destructure+splat.

chrisdickinson commented on services/web/handlers/auth.js:
```diff
@@ -9,23 +9,20 @@ const { text } = require('micro');
+  fork.get('/signup', redirectAuthenticated(signup)),
+  fork.post('/signup', redirectAuthenticated(signupAction)),
+  fork.get('/tokens', seasurf(redirectUnauthenticated(tokens))),
+  fork.post('/tokens', seasurf(redirectUnauthenticated(handleTokenAction)))
```
Ideally this lands at `/settings/:username/tokens`, but I'm going to leave that for another PR.
